purchases/src/main/kotlin/com/revenuecat/purchases/galaxy/handler/ProductDataHandler.kt

```diff
@@ -0,0 +1,182 @@
+package com.revenuecat.purchases.galaxy.handler
+
+import android.os.Handler
+import com.revenuecat.purchases.ProductType
+import com.revenuecat.purchases.PurchasesError
+import com.revenuecat.purchases.PurchasesErrorCallback
+import com.revenuecat.purchases.PurchasesErrorCode
+import com.revenuecat.purchases.common.LogIntent
+import com.revenuecat.purchases.common.StoreProductsCallback
+import com.revenuecat.purchases.common.log
+import com.revenuecat.purchases.galaxy.GalaxyStrings
+import com.revenuecat.purchases.galaxy.IAPHelperProvider
+import com.revenuecat.purchases.galaxy.listener.ProductDataResponseListener
+import com.revenuecat.purchases.galaxy.toStoreProduct
+import com.revenuecat.purchases.galaxy.utils.isError
+import com.revenuecat.purchases.models.StoreProduct
+import com.samsung.android.sdk.iap.lib.vo.ErrorVo
+import com.samsung.android.sdk.iap.lib.vo.ProductVo
+import java.util.ArrayList
+
+internal class ProductDataHandler(
+    private val iapHelper: IAPHelperProvider,
+    private val mainHandler: Handler,
+) : ProductDataResponseListener {
+
+    companion object {
+        private const val GET_PRODUCT_DATA_TIMEOUT_MILLIS = 10_000L
+    }
+
+    @get:Synchronized
+    private var inFlightRequest: Request? = null
+
+    private data class Request(
+        val productIds: Set<String>,
+        val productType: ProductType,
+        val onReceive: StoreProductsCallback,
+        val onError: PurchasesErrorCallback,
+    )
+
+    @get:Synchronized
+    internal val productDataCache = mutableMapOf<String, ProductVo>()
+
+    override fun getProductDetails(
+        productIds: Set<String>,
+        productType: ProductType,
+        onReceive: (List<StoreProduct>) -> Unit,
+        onError: (PurchasesError) -> Unit,
+    ) {
+        if (productIds.isEmpty()) {
+            // Note: This early exit is more than just an optimization. If we call iapHelper.getProductsDetails
+            // with an empty string param (this would happen if productIds is empty), then iapHelper.getProductsDetails
+            // returns all products for the app, which wouldn't give us the product we want (none).
+            log(LogIntent.DEBUG) { GalaxyStrings.EMPTY_GET_PRODUCT_DETAILS_REQUEST }
+            onReceive(emptyList())
+            return
+        }
+
+        if (inFlightRequest != null) {
+            log(LogIntent.GALAXY_ERROR) { GalaxyStrings.ANOTHER_GET_PRODUCT_DETAILS_REQUEST_IN_FLIGHT }
+            val error = PurchasesError(
+                code = PurchasesErrorCode.OperationAlreadyInProgressError,
```

Comment:
Hmm If the parameters matched, we could try to reuse the existing operation... I wonder if we should try to build some kind of queue for these, so we don't fail... But maybe it's an optimization we can work on later.

Reply:
If the `SerialRequestExecutor` is doing it's job properly, this `if` block should never be executed, it's really here as an extra layer of protection 🤞